test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Audience.cs

```diff
@@ -0,0 +1,178 @@
+                        // ValidateTokenAsync with ValidationParameters returns a different error message to account for the
+                        // removal of the ValidAudience property from the ValidationParameters class.
+                        ExpectedExceptionValidationParameters = ExpectedException.SecurityTokenInvalidAudienceException("IDX10215:"),
+                    },
+                    new ValidateTokenAsyncAudienceTheoryData("Valid_AudienceWithinValidAudiences")
+                    {
+                        TokenAudience = Default.Audience,
+                        TVPAudiences = ["ExtraAudience", Default.Audience, "AnotherAudience"],
+                        ValidationParameters = CreateValidationParameters(["ExtraAudience", Default.Audience, "AnotherAudience"]),
+                    },
+                    new ValidateTokenAsyncAudienceTheoryData("Valid_AudienceWithSlash_IgnoreTrailingSlashTrue")
+                    {
+                        // Audience has a trailing slash, but IgnoreTrailingSlashWhenValidatingAudience is true.
+                        TokenAudience = Default.Audience + "/",
+                        TVPAudiences = [Default.Audience],
+                        ignoreTrailingSlashWhenValidatingAudience = true,
+                        ValidationParameters = CreateValidationParameters([Default.Audience], true),
+                    },
+                    new ValidateTokenAsyncAudienceTheoryData("Invalid_AudienceWithSlash_IgnoreTrailingSlashFalse")
+                    {
+                        // Audience has a trailing slash and IgnoreTrailingSlashWhenValidatingAudience is false.
+                        TokenAudience = Default.Audience + "/",
+                        TVPAudiences = [Default.Audience],
+                        ValidationParameters = CreateValidationParameters([Default.Audience], false),
+                        ExpectedIsValid = false,
+                        ExpectedException = ExpectedException.SecurityTokenInvalidAudienceException("IDX10214:"),
+                        ExpectedExceptionValidationParameters = ExpectedException.SecurityTokenInvalidAudienceException("IDX10215:"),
+                    },
+                    new ValidateTokenAsyncAudienceTheoryData("Valid_ValidAudiencesWithSlash_IgnoreTrailingSlashTrue")
+                    {
+                        // ValidAudiences has a trailing slash, but IgnoreTrailingSlashWhenValidatingAudience is true.
+                        TokenAudience = Default.Audience,
+                        ignoreTrailingSlashWhenValidatingAudience = true,
+                        TVPAudiences = [Default.Audience + "/"],
+                        ValidationParameters = CreateValidationParameters([Default.Audience + "/"], true),
+                    },
+                    new ValidateTokenAsyncAudienceTheoryData("Invalid_ValidAudiencesWithSlash_IgnoreTrailingSlashFalse")
```

Comment:
maybe I missed it, but do we need tests around:
Empty Audiences:
 - test with an empty audience string in the token and validation parameters.
 - test with null as the token audience
 
Case Sensitivity
Null Validation Parameters
Multiple Valid Audiences
 - overlapping valid audience where one is a substring of another

Reply:
Since we have the model we're implementing this time around as opposed to when we began working on JsonWebTokenHandler we already tested that as part of this set of [tests](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/test/Microsoft.IdentityModel.Tokens.Tests/Validation/AudienceValidationResultTests.cs) because our handlers will call the same validators underneath the hood, we can go directly into integration and regression tests. We could add more cases later on but the focus of this PR is to be able to match the progress we're doing on JsonWebTokenHandler.

Reply:
Maybe we have a gap with JsonWebTokenHandler?

Reply:
I think in this case it's unlikely since, we have both unit and functional tests for each validator already, so we ensure the validation part it's covered on those. Mainly we're looking for gaps related to the experience of the users in these PR's that will compare the old call path vs the new one.